flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -391,6 +401,10 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
                         () -> throwException(metadata, exception, producer),
                         "Failed to send data to Kafka");
             }
+
+            for (Consumer<RecordMetadata> metadataConsumer : metadataConsumers) {
+                metadataConsumer.accept(metadata);
```

Comment:
Is this really safe? Currently one of the Kafka producer threads will update the metadata.

Reply:
Good point, I've summarized below